flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -860,6 +902,7 @@ private ExecutionGraph createExecutionGraphAndRestoreState(JobGraph adjustedJobG
                 TaskDeploymentDescriptorFactory.PartitionLocationConstraint.MUST_BE_KNOWN,
                 initializationTimestamp,
                 vertexAttemptNumberStore,
+                computeVertexParallelismStore(adjustedJobGraph),
```

Comment:
hmm...if we compute the maxParallelism here, couldn't this result in us using different maxParallelism values for different execution attempts, and thus by extension, savepoints/checkpoints? Because the adjustedJobGraph may have a different parallelism each time.
I think this is a problem because it turns loading a savepoint/checkpoint from this job into a guessing game of "what-max-parallelism-was-set-this-time".
IOW, on every scale up the job may suddenly break, if the maxP ends up being higher than in the checkpoint we restore from.

We can't base it on the original JobGraph, because there the parallelism is cranked to the max and we wouldn't be able to restore from a savepoint (and tank performance significantly).
We also can't base it on the first execution, because in HA on JM failover we could again end up with a different maxParallelism...
We may just have to force users to set the maxParallelism when using reactive mode.

Reply:
Actually, if we were to base the maxP on the original-original JobGraph (the one in the DefaultJobManagerRunnerFactory), then that would also work, and be most in line with existing behavior.

Reply:
Very nice catch, I had not considered how this fits into the declarative slot pool at all. By using the original-original JobGraph that's passed into the AdaptiveScheduler to compute the VertexParallelismStore immediately, it makes it quite easy to integrate into that resource declaration computation. I just had to use the newly created parallelism store as the source of truth in JobGraphJobInformation's parallelism reporting. 👍🏼 Thanks very much Chesnay

Reply:
This was easy to integrate, but it might show some issues with the `SlotSharingSlotAllocator#tryReserveResources` method – I'm not entirely sure where they lie/ if it is actually an issue with the parallelism assignment. It could also be an issue with the testing setups in AdaptiveSchedulerTest, which make some assumptions about parallelism. Investigating...